Optimal transport

rousseab opened the pull request:
Here I take a first go at using "optimal transport" to regularize the relative coordinate score.

One big change is that I now push the "noising" of the training data to the data loader instead of doing it in the lightning module. This is now the responsibility of the `NoisingTransform` class.

This is necessary because the linear assignment problem (LAP) (ie, the hungarian algorithm) only works on CPU.  Irrespective of whether we end up using optimal transport, I think it's still a good idea to push this noising stuff to "data preprocessing" instead of the main training loop. It does remove some crud from the Lightning module, which no longer needs to know how to create noise.

--> This does mean, however, that the dataloader config must now be provided with `NoiseParameters`. This means that configs must be jiggled a bit.

The optimal transport stuff is the responsibility of the `Transporter` class, which implements a heuristic to find the best symmetry group operations to align two configurations. We might come up with a better heuristic in the future, I'm pretty sure it's not optimal.

I also implement an equivariant analytical score network based on this optimal transport stuff. It is useful to test ideas / benchmark the sampling code!


Finally, I initially relied on the POT library to solve the LAP problem. After some device bjorks, I decided to just use the scipy algorithm instead. I did jiggle the uv.lock file a bit because of this (adding POT, removing POT...): I don't think the changes there are important or meaningful.

sblackburn86 commented on src/diffusion_for_multi_scale_molecular_dynamics/data/diffusion/noising_transform.py:
```diff
@@ -0,0 +1,177 @@
+
+        if self.use_optimal_transport:
+            # TODO: review this as we improve the transporter
+            self.point_group_operations = torch.diag(
```
torch.diag(torch.ones()) = torch.eye()

rousseab replied:
I fixed this.

sblackburn86 reviewed (approved):
I made a comment for a function substitution suggestion, but it's optional. Code looks good. You changed the uv.lock - I assume it's for POT. This should be reflected in the TOML too

rousseab commented:
I reverted the uv.lock file to what it is in the main branch. During development, I added POT, and then I removed POT. This changed the uv.lock file (I guess it updated other things unrelated to POT). I remove these changes in the interest of clarity.
